gamemodes/terrortown/gamemode/server/sv_roleselection.lua

```diff
@@ -0,0 +1,496 @@
+		if maxBaseroles == 0 then
+			maxBaseroles = nil
+		end
+	end
+
+	local curRoles = 2 -- amount of roles, start with 2 because INNOCENT and TRAITOR are all the time available
+	local curBaseroles = 2 -- amount of base roles, ...
+
+	-- we have to create a enumerable table to get random results easily
+	local availableBaseRolesTbl = {}
+	local availableSubRolesTbl = {}
+	local availableBaseRolesAmount = 0
+	local availableSubRolesAmount = 0
+
+	for roleData in pairs(rolesAmountList) do
+		if roleData.baserole then
+			availableSubRolesAmount = availableSubRolesAmount + 1
+
+			availableSubRolesTbl[availableSubRolesAmount] = roleData
+		else
+			availableBaseRolesAmount = availableBaseRolesAmount + 1
+
+			availableBaseRolesTbl[availableBaseRolesAmount] = roleData
+		end
+	end
+
+	local selectableRoles = {}
+
+	-- first of all, we need to select the baseroles. Otherwise, we would select subroles that never gonna be choosen because if the missing baserole
+	for i = 1, availableBaseRolesAmount do
+		if maxRoles and maxRoles <= curRoles or maxBaseroles and maxBaseroles <= curBaseroles then break end -- if the limit is reached, stop selection
+
+		local rnd = math.random(#availableBaseRolesTbl)
+		local roleData = availableBaseRolesTbl[rnd]
+
+		table.remove(availableBaseRolesTbl, rnd) -- selected roleData shouldn't get selected multiple times
+
+		selectableRoles[roleData] = rolesAmountList[roleData]
+
+		curRoles = curRoles + 1
+		curBaseroles = curBaseroles + 1
+	end
+
+	-- now we need to select the subroles
+	for i = 1, availableSubRolesAmount do
+		if maxRoles and maxRoles <= curRoles then break end -- if the limit is reached, stop selection
+
+		local rnd = math.random(#availableSubRolesTbl)
+		local roleData = availableSubRolesTbl[rnd]
+
+		table.remove(availableSubRolesTbl, rnd) -- selected roleData shouldn't get selected multiple times
+
+		selectableRoles[roleData] = rolesAmountList[roleData]
+
+		curRoles = curRoles + 1
+	end
+
+	hook.Run("TTT2ModifySelectableRoles", selectableRoles)
+
+	roleselection.selectableRoles = selectableRoles
+
+	return selectableRoles
+end
+
+---
+-- if possible, give plys the available roles
+local function SetSubRoles(plys, prevRoles, availableRoles, selectableRoles)
+	local plysAmount = #plys
+	local availableRolesAmount = #availableRoles
+	local tmpSelectableRoles = table.Copy(selectableRoles)
+
+	while plysAmount > 0 and availableRolesAmount > 0 do
+		local pick = math.random(plysAmount)
+		local ply = plys[pick]
+
+		local rolePick = math.random(availableRolesAmount)
+		local roleData = availableRoles[rolePick]
+		local roleCount = tmpSelectableRoles[roleData]
+
+		local minKarmaCVar = GetConVar("ttt_" .. roleData.name .. "_karma_min")
+		local minKarma = minKarmaCVar and minKarmaCVar:GetInt() or 0
+
+		-- give this player the role if
+		if plysAmount <= roleCount -- or there aren't enough players anymore to have a greater role variety
```

Comment:
Shouldn't this be `plysAmount < roleCount` ? Because there is still enough space to fit all roles.